src/datacite/datacite.py

```diff
@@ -0,0 +1,311 @@
+        ...     result = fetch_doi_page(
+        ...         session=session,
+        ...         prefixes=["10.7303"],
+        ...         state="findable",
+        ...         page_size=100,
+        ...         page_number=0,
+        ...         detail=True
+        ...     )
+        ...     print(f"Got {len(result['data'])} DOIs")
+
+Notes:
+    - DataCite provides higher rate limits (1000 requests per 5 minutes) when
+      you include an email address in the User-Agent header via user_agent_mailto.
+    - The API uses zero-based pagination.
+    - Maximum page size is 1000 items.
+    - All `fetch_*` functions use exponential backoff for retryable errors.
+"""
+import gzip
+import json
+import time
+from typing import Iterable, Dict, Any, Optional, List
+
+import requests
+
+DATACITE_API = "https://api.datacite.org/dois"
+
+
+def _build_query_params(
+    prefixes: List[str],
+    state: str,
+    page_size: int,
+    page_number: int,
+    detail: bool,
+) -> Dict[str, Any]:
+    """Build query parameters for DataCite API request.
+
+    Args:
+        prefixes: List of DataCite DOI prefixes to query (e.g., ["10.7303"]).
+        state: DOI state to filter by ("findable", "registered", or "draft").
+        page_size: Number of results to return per page.
+        page_number: Zero-based page number for pagination.
+        detail: Whether to include detailed metadata in the response.
+
+    Returns:
+        Dictionary of query parameters formatted for the DataCite API.
+    """
+    params = {
+        "prefix": ",".join(prefixes),
+        "state": state,
+        "page[size]": page_size,
+        # Design doc example shows 0-based; we follow it.
+        "page[number]": page_number,
+    }
+    if detail:
+        params["detail"] = "true"
+    return params
+
+
+def _make_request_with_retry(
+    session: requests.Session,
+    url: str,
+    params: Dict[str, Any],
+    timeout: int = 60,
+    initial_backoff: int = 2,
+    max_backoff: int = 60,
+) -> requests.Response:
+    """Make HTTP GET request with exponential backoff for retryable errors.
+
+    Automatically retries requests on rate limiting (429) and server errors
+    (500, 502, 503, 504) using exponential backoff strategy.
+
+    Args:
+        session: Requests session object for making HTTP calls.
+        url: URL to send the GET request to.
+        params: Query parameters to include in the request.
+        timeout: Request timeout in seconds. Defaults to 60.
+        initial_backoff: Initial backoff delay in seconds. Defaults to 2.
+        max_backoff: Maximum backoff delay in seconds. Defaults to 60.
+
+    Returns:
+        Successful HTTP response object with status code 200.
+
+    Raises:
+        requests.HTTPError: For non-retryable HTTP errors (e.g., 400, 401, 404).
+    """
+    backoff = initial_backoff
+    while True:
+        resp = session.get(url, params=params, timeout=timeout)
+        if resp.status_code == 200:
+            return resp
```

Comment:
Might also need to handle for a few more status code, like a `204 No Content` response.